### PR TITLE
[Perf] Skip invisible Hopper FP4 indexer tiles

### PR DESCRIPTION
**Pull request by ZenAlexa**
## Motivation

The Hopper FP4 indexer in #38798 launches against the static replay capacity. Device-visible lengths can cover a small prefix of that capacity. Every allocated tile currently performs FP4 decoding, query loads and two dot products.

## Modifications

I skip the score computation for tiles whose first position is outside the current device-visible range. Both branches write their output directly, including the negative-infinity tail. Partial tiles keep the existing masks, arithmetic and bf16 rounding points. The wrapper, output capacity and downstream top-k interface stay unchanged.

This targets `dsv4.1` and is independent of #38848.

## Accuracy Tests

I ran the production kernel and the registered test assertions on H100 with PyTorch 2.13.0+cu130 and Triton 3.7.1. Zero-tolerance reference checks cover H16 and H32, capacities 0, 63, 64, 65 and 129, mixed visible lengths and reordered physical pages. Fixed-address CUDA graph replays cover full capacity, zero, 65, 63 and full capacity again.

All logits and Top-512 indices exactly match the baseline in 25 additional large-capacity cases. Full-package test collection and end-to-end serving remain pending.

I merged the refreshed `dsv4.1@a186571c17b15dd5627e57257c1b2696959db01e` into this branch, preserving the existing contribution history. The diff remains limited to the kernel and its registered test in the current taxonomy. The kernel and test contents match the measured revision. The registered-test linter and suite discovery pass at the updated head. The test command is:

```bash
python test/registered/kernel/ops/attention/test_sm90_fp4_indexer.py
```

### Reported serving result

[Orien-Xu reported](https://github.com/sgl-project/sglang/pull/39046#issuecomment-5641748979) a four-H200-SXM run of DeepSeek-V4.1-Flash with TP4, EP4, a 262,144-token context capacity, `1k` input tokens, `1k` output tokens, closed-loop requests, and speculation disabled. Their revision is `dsv4.1@da64c5cbb` plus an equivalent early-exit patch. At concurrency 32, TPOT changed from 42.3 ms to 31.6 ms and aggregate throughput from 713 to 934 tokens per second. They reported unchanged concurrency-1 performance and identical greedy outputs. The configuration and revision above are preserved from the linked report.

#### Trace-derived profiling follow-up

I parsed the two rank-0 traces in [Orien-Xu's follow-up](https://github.com/sgl-project/sglang/pull/39046#issuecomment-5647908377). Each contains 24 `cudaGraphLaunch` replays with eight `_fp4_index_logits_kernel` launches per replay. Their reported setup uses four H200 SXM GPUs, TP4, EP4, GPU-resident Engram tables, speculation disabled, and `--context-length 262144 --cuda-graph-max-bs 32 --mem-fraction-static 0.93`. The profiling load has 32 concurrent requests with approximately 1.2k input tokens and live lengths of 1.2k–1.8k tokens. They identify the tree as `dsv4.1@da64c5cbb` integrated into their fork.

| Trace-derived metric | Original kernel | Early-exit kernel |
|---|---|---|
| Median summed indexer duration per replay | 11.546 ms | 0.583 ms |
| Median summed graph-kernel duration per replay | 39.822 ms | 28.949 ms |
| Median indexer launch duration, grid `(32, 2049, 1)` | 891.746 µs | 45.152 µs |
| Median indexer launch duration, grid `(32, 4097, 1)` | 1773.380 µs | 89.440 µs |

Per-replay sums group kernel events by their `cudaGraphLaunch` correlation ID, followed by the median across 24 replays. Launch-duration medians use 72 events for the smaller grid and 120 for the larger grid. Both traces retain these grids. Summed kernel durations count overlapping intervals separately and cover rank 0. TPOT and aggregate throughput are the reporter's separate serving measurements above.

## Speed Tests and Profiling

NVIDIA H100 80GB HBM3, CUDA 13.0, PyTorch 2.13.0+cu130, Triton 3.7.1, H32, page size 64. Baseline is `3b709e55c0f7599f90bdd400e1fe758c5a942cb6`. Each row uses ten CUDA-event samples of thirty graph replays, alternating variant order. These timings cover `fp4_index_logits_decode`.

| Batch | Capacity | Visible positions | Baseline µs | This change µs | Speedup |
|---|---|---|---|---|---|
| 1 | 65,536 | 4096 | 17.821 | 6.457 | 2.76x |
| 4 | 262,144 | 4096 | 226.079 | 14.226 | 15.89x |
| 4 | 1,048,576 | 64 | 846.802 | 41.595 | 20.36x |
| 4 | 1,048,576 | 4096 | 847.926 | 43.616 | 19.44x |
| 4 | 1,048,576 | [0, 63, 4096, 1048576] | 848.245 | 246.852 | 3.44x |

Full-capacity measurements include the branch cost.

| Batch | Full capacity | Baseline µs | This change µs | Latency change |
|---|---|---|---|---|
| 1 | 65,536 | 17.904 | 18.477 | +3.20% |
| 1 | 262,144 | 60.684 | 61.082 | +0.66% |
| 1 | 1,048,576 | 222.354 | 225.193 | +1.28% |
| 4 | 65,536 | 60.762 | 61.157 | +0.65% |
| 4 | 262,144 | 223.985 | 222.756 | -0.55% |
| 4 | 1,048,576 | 865.138 | 874.128 | +1.04% |

The largest full-capacity percentage increase is 3.20%, adding 0.573 µs at batch 1 and capacity 65,536. At batch 4 and capacity 1,048,576, the increase is 1.04%. Output allocation size remains unchanged.

<details>
<summary>Reproduce the kernel timing from this branch</summary>

Set `batch`, `capacity` and `visible` to the desired row.

```python
import importlib.util
import statistics
import subprocess
import tempfile
from pathlib import Path
import torch

kernel = "python/sglang/kernels/ops/attention/dsv4/sm90_fp4_indexer.py"
base = "3b709e55c0f7599f90bdd400e1fe758c5a942cb6"
with tempfile.TemporaryDirectory() as directory:
    original = Path(directory) / "baseline.py"
    original.write_bytes(subprocess.check_output(["git", "show", f"{base}:{kernel}"]))
    functions = []
    for name, path in (("baseline", original), ("candidate", Path(kernel))):
        spec = importlib.util.spec_from_file_location(name, path)
        module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(module)
        functions.append(module.fp4_index_logits_decode)
    torch.manual_seed(42)
    batch, capacity, visible = 4, 1048576, 4096
    table = torch.randint(0, 256, (capacity // 64, 4352), device="cuda", dtype=torch.uint8)
    table[:, 4096:] = torch.randint(125, 129, (capacity // 64, 256), device="cuda", dtype=torch.uint8)
    q = torch.randn(batch, 32, 128, device="cuda", dtype=torch.bfloat16)
    weights = torch.randn(batch, 32, device="cuda", dtype=torch.bfloat16)
    slots = torch.arange(capacity, device="cuda").expand(batch, -1).contiguous()
    lens = torch.tensor([visible] * batch if isinstance(visible, int) else visible, device="cuda", dtype=torch.int64)
    inputs = (q, weights, slots, lens, table, 64)
    graphs, outputs = [], []
    for fn in functions:
        for _ in range(3):
            fn(*inputs)
        graph = torch.cuda.CUDAGraph()
        with torch.cuda.graph(graph):
            outputs.append(fn(*inputs))
        graphs.append(graph)
    for graph in graphs:
        for _ in range(10):
            graph.replay()
    assert torch.equal(*outputs)
    samples = [[], []]
    for sample in range(10):
        for index in ((0, 1) if sample % 2 == 0 else (1, 0)):
            start = torch.cuda.Event(enable_timing=True)
            end = torch.cuda.Event(enable_timing=True)
            start.record()
            for _ in range(30):
                graphs[index].replay()
            end.record()
            end.synchronize()
            samples[index].append(start.elapsed_time(end) * 1000 / 30)
    print(torch.__version__, torch.version.cuda, torch.cuda.get_device_name())
    print("baseline us", statistics.median(samples[0]))
    print("candidate us", statistics.median(samples[1]))
```
</details>

## Checklist

- [x] Format the changed files with the project pre-commit hooks.
- [x] Add registered kernel tests.
- [x] Keep the existing public interface and documentation contract.
- [x] Provide kernel accuracy and speed benchmark results.
- [x] Follow the SGLang code style.

---
### CI States

Latest PR Test (Base): :x: [Run #35048249276](https://github.com/sgl-project/sglang/actions/runs/35048249276)
Latest PR Test (Extra): :x: [Run #35048249006](https://github.com/sgl-project/sglang/actions/runs/35048249006)
Latest PR Test (AMD ROCm 10): :x: [Run #35048249266](https://github.com/sgl-project/sglang/actions/runs/35048249266)

### CODE REVIEW

**Comment by ZenAlexa** [comment]
@hnyls2002 I kept this based on `dsv4.1`, where the Hopper kernel lives. The CI gate requires `main` commit `3700c4ee26a1`; could you help run the registered H100 test against this branch?

**Comment by Orien-Xu** [comment]
Add a confirmation here on our 4x H200 SXM (TP4/EP4, DeepSeek-V4.1-Flash, dsv4.1 @da64c5cbb + an equivalent early-exit) with --context-length 262144, 1k/1k closed loop, no speculation, C=32 TPOT 42.3 → 31.6 ms and 713 → 934 tok/s aggregate; C=1 unchanged; greedy outputs identical.

Before the change the decode indexer logits kernel was 28.9% of step GPU time at bs=32 (1.43 ms/call × 8 index layers). Happy to share the profiles

**Comment by ZenAlexa** [comment]
Thanks for checking this on H200. I'd like to see the C=32 decode profiles with and without the early exit, especially the indexer time across those eight layers.

Could you share those traces and the small patch you tested? That would help me connect the serving result to this kernel change.

**Comment by Orien-Xu** [comment]
Lost the exact traces but here is another run.

Traces attached (rank 0, torch profiler, GPU activities only, 24 decode steps at batch 32, 4x H200 SXM TP4/EP4,
`--context-length 262144 --cuda-graph-max-bs 32 --mem-fraction-static 0.93`, no speculative decoding, Engram
tables on the GPUs). Same tree for both runs (upstream dsv4.1 @da64c5cbb on top of our fork main); the only
difference is the attached 10-line patch to `sm90_fp4_indexer.py`. Load: 32 concurrent requests of ~1.2k prompt
tokens, so each request's live length is 1.2k-1.8k tokens against the 262144-token reserved window.

|                                   | no early exit | early exit |
|-----------------------------------|---------------|------------|
| GPU kernel time per decode step   | 39.99 ms      | 29.09 ms   |
| `_fp4_index_logits_kernel`/step   | 11.55 ms (28.9%) | 0.58 ms (2.0%) |
| per launch, index layers 1-3      | 892 us        | 45 us      |
| per launch, index layers 4-8      | 1773 us       | 89 us      |

(8 launches per step; layers 1-3 have half the reserved width of layers 4-8, hence the 2x. Medians over 24 steps,
min/max within 1%.) Serving effect of the same change at C=32, 1k in / 1k out: TPOT 42.3 -> 31.6 ms
(713 -> 934 tok/s aggregate); greedy outputs identical, since the skipped tiles were already masked to -inf.

attaching two trace file
[exit-bs32-TP-0-EP-0.trace.json.gz](https://github.com/user-attachments/files/32148400/exit-bs32-TP-0-EP-0.trace.json.gz)
[noexit-bs32-TP-0-EP-0.trace.json.gz](https://github.com/user-attachments/files/32148402/noexit-bs32-TP-0-EP-0.trace.json.gz)

perfectto here for convenience
with early exit
annotated-exit-45us-grid

without early exit
annotated-noexit-890us-grid

**Comment by ZenAlexa** [comment]
Thanks for sharing these, I've added the profiles to the PR.

**Comment by simple-sun** [comment]
Hi @ZenAlexa, this PR appears to have been closed when the dsv4.1 base branch was deleted. The branch is now available again, and #38798 is still open. Would it be possible to reopen this PR?

We independently tested an equivalent tile-skipping change on H20 GPUs with DeepSeek-V4.1-Flash, TP4/EP4, DSPARK In a preliminary 1P1D test using the same 180 requests at 35 RPM, mean TPOT decreased from 17.84 ms to 7.38 ms, with all requests succeeding in both runs.

These are preliminary results: generated output lengths differed between runs, so the full end-to-end improvement cannot be attributed solely to the kernel change. Kernel-level checks passed for boundary lengths, randomized slots, and CUDA Graph replay with changing visible lengths. Longer serving tests and full-context validation are still pending.

Happy to contribute our H20 + DSPARK validation results to this PR.

**Comment by ZenAlexa** [comment]
I've confirmed the PR is open again on `dsv4.1`. Please add your H20 + DSPARK results here; the full-context checks and a serving run with matched output lengths would be especially useful.
